### PR TITLE
196 user recommendations

### PR DESCRIPTION
**Pull request by ishootdice**
- Configured neo4j connection to project
- Added recomendations queue for RabbitMQ
- Implemented data access and recommendation services
- Defined queries for neo4j database
- Recommendation service added to User service
- Added Endpoint to Recommendation Controller to retrieve user recommendations
- Added httpClient inside of User service to be able to sent request on RecommendationController endpoint
- Configured docker-compose files

### CODE REVIEW

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs [hunk 1]
```diff
@@ -94,6 +115,13 @@ public async Task<List<UserShortInfoDto>> GetFilteredUsers(UserFilterDto userFil
         {
             filteredUsers = filteredUsers.Where(u => u.Tags.Any(t => filter.Topics.Contains(t.Name)));
         }
+        if (filter.Compatibility is not null)
+        {
+            var compatibleUsers =
+                await GetRecommendedUsers((int)filter.Compatibility, filteredUsers.Select(x => x.Id).ToList());
+            filteredUsers = filteredUsers.Where(x => compatibleUsers.Contains(x.Id));
+        }
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs:123]: Have you set some value to recomendation to show it on FE?

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs [hunk 2]
```diff
@@ -123,18 +151,27 @@ public async Task<LessonDto> EnrollUserToLesson(long lessonId)
...
+        {
+            dto.SubscribersCount = _context.Lessons
+                .Select(t => new {Id = t.Id, SbCount = t.Subscribers.Count})
+                .FirstOrDefault(l => l.Id == lessonId)!.SbCount;
+        }
 
-        return _mapper.Map<Lesson, LessonDto>(lesson, options => options.AfterMap(AfterMapAction));
+        var lessonDto = _mapper.Map<Lesson, LessonDto>(lesson, options => options.AfterMap(AfterMapAction));
+        
+        SendAddClassQuery(_authService.UserId, lessonDto);
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs:172]: Isn't it better to make it after lesson is started? As it may be cancelled and this values won't make any sense...

File: backend/EasySpeak.Core/EasySpeak.Core.Common/DTO/Rabbit/RecommendationServiceMessageDto.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace EasySpeak.Core.Common.DTO.Rabbit;
+
+public class RecommendationServiceMessageDto
+{
+    public QueryType Type { get; set; }
+    public Dictionary<string, object> Parameters { get; set; } = new();
+    
+    [JsonConverter(typeof(ListStringConverter))]
+    public List<string>? ParameterList { get; set; } = new();
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.Common/DTO/Rabbit/RecommendationServiceMessageDto.cs:13]: Use string[], it should work without converters

File: backend/EasySpeak.Core/EasySpeak.Core.Common/Enums/QueryType.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+namespace EasySpeak.Core.Common.Enums;
+
+public enum QueryType
+{
+    AddUser,
+    AddTags,
+    AddClass,
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.Common/Enums/QueryType.cs:7]: Maybe AddClass -> StartClass?

File: backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Extensions/ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -25,19 +25,25 @@ public static void RegisterCustomServices(this IServiceCollection services, ICon
...
+                new ConnectionProvider(configuration.GetValue<string>("Rabbit")));
             services.AddTransient<IMessageProducer, MessageProducer>();
             services.AddTransient<IHttpRequestService, HttpRequestService>();
             services.AddScoped<IFirebaseAuthService, FirebaseAuthService>();
-            services.AddScoped<IUserService, UserService>();
             services.AddScoped<IChatService, ChatService>();
             services.AddFirebaseApp();
-            services.AddScoped<INotificationService, NotificationService>();
+            var test = new RabbitQueuesOptions();
+            configuration.GetSection("RabbitQueues").Bind(test);
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Extensions/ServiceCollectionExtensions.cs:42]: Do you need this?

File: backend/EasySpeak.Notifier/EasySpeak.Notifier.WebAPI/Services/ConsumerHostedService.cs [hunk 6]
```diff
@@ -34,11 +34,11 @@ protected override Task ExecuteAsync(CancellationToken stoppingToken)
 
             try
             {
-                _consumer.Recieve<Tuple<string, NotificationDto>>((data) =>
+                _consumer.Receive<Tuple<string, NotificationDto>>((data) =>
                 {
                     if (data is not null)
                     {
-                        _hubContext.Clients.User(data.Item1).SendAsync("Notify", JsonConvert.SerializeObject(data.Item2));
+                        _hubContext.Clients.User(data.Item1).SendAsync("Notify", JsonConvert.SerializeObject(data.Item2),  stoppingToken);
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Notifier/EasySpeak.Notifier.WebAPI/Services/ConsumerHostedService.cs:41]: stoppingToken -> cancellationToken or ct

File: backend/EasySpeak.RecommendationService/.idea/.idea.EasySpeak.RecommendationService/.idea/workspace.xml
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.RecommendationService/.idea/.idea.EasySpeak.RecommendationService/.idea/workspace.xml:1]: Can you add it to gitignore, please?

File: backend/EasySpeak.RecommendationService/.DS_Store
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.RecommendationService/.DS_Store:1]: Is this file needed?

File: backend/EasySpeak.RecommendationService/EasySpeak.RS.WebAPI/Controllers/RecommendationController.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+    public RecommendationController(IRecommendationService recommendationService)
+    {
+        _recommendationService = recommendationService;
+    }
+
+    [HttpPost]
+    public async Task<List<long>> GetRecommendedUsers(NewRecommendationDto recommendationDto)
+    {
+        return await _recommendationService.GetRecommendedUsers(recommendationDto.ToDictionary());
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.RecommendationService/EasySpeak.RS.WebAPI/Controllers/RecommendationController.cs:21]: Again, can you return at least some recommendation score?
And when we don't have recommendation filter, it's also better to set some scores...

File: backend/EasySpeak.RecommendationService/EasySpeak.RS.WebAPI/Extensions/ServiceCollectionExtentions.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace EasySpeak.RS.WebAPI.Extensions;
+
+public static class ServiceCollectionExtensions
+{
+    public static void RegisterCustomServices(this IServiceCollection services,
+        IConfiguration configuration)
+    {
+        var serviceOption = new RecommendationServiceOptions();
+        configuration.GetSection("Neo4jSettings")
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.RecommendationService/EasySpeak.RS.WebAPI/Extensions/ServiceCollectionExtentions.cs:18]: You get configuration.GetValue<Recommedn....>("Neo4jSettings")

File: backend/EasySpeak.RecommendationService/EasySpeak.RS.WebAPI/Services/ConsumerHostedService.cs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+    private async Task SendQuery(RecommendationServiceMessageDto data)
+    {
+        switch (data.Type)
+        {
+            case QueryType.AddClass:
+                await _recommendationService.AddClass(data.Parameters);
+                break;
+            case QueryType.AddTags:
+                if(data.ParameterList is not null) await _recommendationService.AddTags(data.Parameters, data.ParameterList);
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.RecommendationService/EasySpeak.RS.WebAPI/Services/ConsumerHostedService.cs:63]: if ( is null) break;

File: docker-compose.services.yml
Comment (Anatolii-Ivanov) [inline docker-compose.services.yml:1]: Ask Tania to check volumes, please, as she may store something in that folders before merge

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs (hunk 1 above)
Comment (ishootdice) [reply backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs:123]: Yes, It works, but I forgot to make screenshot, I will do it with requested changes

File: backend/EasySpeak.Core/EasySpeak.Core.Common/DTO/Rabbit/RecommendationServiceMessageDto.cs (hunk 3 above)
Comment (ishootdice) [reply backend/EasySpeak.Core/EasySpeak.Core.Common/DTO/Rabbit/RecommendationServiceMessageDto.cs:13]: I've checked how it works with array, and result is the same. Anyway, I will check it again, and if it works fine, I'll change it

File: backend/EasySpeak.Core/EasySpeak.Core.Common/Enums/QueryType.cs (hunk 4 above)
Comment (ishootdice) [reply backend/EasySpeak.Core/EasySpeak.Core.Common/Enums/QueryType.cs:7]: Ok, I will correct the naming

File: backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Extensions/ServiceCollectionExtensions.cs (hunk 5 above)
Comment (ishootdice) [reply backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Extensions/ServiceCollectionExtensions.cs:42]: Yeah, If I'm not wrong, I can't register GraphDatabase without passing info from this section, I've tried to do this, but anyway, I'll check it out

Comment (ishootdice) [reply backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Extensions/ServiceCollectionExtensions.cs:42]: No, I definitely don't need this)

**Comment by ishootdice** [comment]
image

**Comment by ishootdice** [comment]
image

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonBackgroundService.cs [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        while (!cancellationToken.IsCancellationRequested)
+        {
+            using (var scope = _services.CreateScope())
+            {
+                var context = scope.ServiceProvider.GetRequiredService<EasySpeakCoreContext>();
+                
+                try
+                {
+                    IAsyncEnumerable<Lesson> lessons = QueryableExtensions.Include(context.Lessons, l => l.Subscribers)
+                        .Where(l => l.IsCalculated == false && l.StartAt >= DateTime.Now && l.StartAt <= DateTime.Now.AddMinutes(30)).AsAsyncEnumerable();
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonBackgroundService.cs:42]: !l.IsCalculated

Comment (Anatolii-Ivanov) [reply backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonBackgroundService.cs:42]: l.StartAt <= DateTIme.Now

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonBackgroundService.cs [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+                    await foreach (var lesson in lessons)
+                    {
+                        if (lesson.Subscribers.Any()) SendLessonStartQuery(lesson);
+                        lesson.IsCalculated = true;
+                    }
+                    
+                    await context.SaveChangesAsync(cancellationToken);
+                    await Task.Delay(TimeSpan.FromMinutes(30), cancellationToken);
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonBackgroundService.cs:51]: Move 30 to appsettings

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonBackgroundService.cs [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+                }
+                catch (Exception ex)
+                {
+                    Console.WriteLine(ex.Message);
+                }
+            }
+        }
+        
+        await Task.Delay(TimeSpan.FromMinutes(30), cancellationToken);
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonBackgroundService.cs:60]: It's not needed, if cancellation is requested, it won't work

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs [hunk 13]
```diff
@@ -94,8 +116,23 @@ public async Task<List<UserShortInfoDto>> GetFilteredUsers(UserFilterDto userFil
...
+            await GetRecommendedUsers(filter.Compatibility, filteredUsersList.Select(x => x.Id).ToList());
+        
+        var compatibleUsers = filteredUsersList.Where(x => compatabilityInformation.ContainsKey(x.Id)).Select(x => x).ToList();
+        
+        var result = _mapper.Map<List<UserShortInfoDto>>(compatibleUsers);
+
+        foreach (var user in result)
+        {
+            user.Compatibility = compatabilityInformation[user.Id];
+        }
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs:133]: Move to separate function like "AddCompatibility"

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs (hunk 2 above)
Comment (Anatolii-Ivanov) [reply backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs:181]: I guess it's not needed anymore

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs [hunk 14]
```diff
@@ -223,9 +269,70 @@ public async Task<UserDto> MakeAdminAsync(int userId)
...
+        queryParams.Parameters.Add("id", id);
+
+        SendQueryToRabbit(queryParams);
+    }
+    
+    private void SendQueryToRabbit(RecommendationServiceMessageDto data)
+    {
+        _messageProducer.Init(_rabbitQueues.RecommendationQueue, "");
+        _messageProducer.SendMessage(data);
+    }
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs:315]: Maybe this part to separate service?

File: backend/EasySpeak.RecommendationService/.DS_Store
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.RecommendationService/.DS_Store:1]: Still this files
[git delk](https://www.git-tower.com/learn/git/commands/git-rm)

File: backend/EasySpeak.RecommendationService/.idea/.idea.EasySpeak.RecommendationService/.idea/workspace.xml
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.RecommendationService/.idea/.idea.EasySpeak.RecommendationService/.idea/workspace.xml:1]: And this

**Review by Anatolii-Ivanov — CHANGES_REQUESTED** [review]

File: backend/EasySpeak.RecommendationService/EasySpeak.RS.WebAPI/Program.cs [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+builder.Services.AddControllers();
+builder.Services.RegisterCustomServices(builder.Configuration);
+// Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
+builder.Services.AddEndpointsApiExplorer();
+builder.Services.AddSwaggerGen();
+builder.Services.AddHealthChecks();
+builder.Services.AddRouting(options => options.LowercaseUrls = true);
+builder.Services.AddHostedService<ConsumerHostedService>();
+builder.WebHost.UseUrls("http://*:5110");
```
Comment (tatianahutii) [inline backend/EasySpeak.RecommendationService/EasySpeak.RS.WebAPI/Program.cs:22]: Add to readme

File: backend/EasySpeak.RecommendationService/EasySpeak.RS.WebAPI/appsettings.json [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+  "AllowedHosts": "*",
+  "Rabbit": "amqp://guest:guest@localhost:5672",
+  "Queues": {
+    "ProjectOutputQueue": {
+      "ExchangeName": "NotifierExchange",
+      "ExchangeType": "direct",
+      "QueueName": "ProjectOutputQueue",
+      "RoutingKey": "projectOutput"
+    }
+  },
```
Comment (tatianahutii) [inline backend/EasySpeak.RecommendationService/EasySpeak.RS.WebAPI/appsettings.json:17]: Are you sure you need this? This code looks like a part of the Codi =)

File: docker-compose.apps.yml [hunk 17]
```diff
@@ -45,6 +45,25 @@ services:
           - back
       environment:
           ASPNETCORE_ENVIRONMENT: 'Development'
+          RecommendationService_Connection: 'bolt://localhost:7687'
+          RecommendationService_User: 'neo4j'
+          RecommendationService_Password: 'e3hfuTFgkUYa4WanMbnlZUseHmmBtC1q'
+          RecommendationService_Database: 'neo4j'
+
+  easyspeak_recommendationservice:
```
Comment (tatianahutii) [inline docker-compose.apps.yml:53]:
```suggestion
  easyspeak_recommendator_service:
```

Comment (tatianahutii) [reply docker-compose.apps.yml:53]: And add to the dependencies for frontend

File: frontend/src/app/shared/utils/user-card.utils.ts [hunk 18]
```diff
@@ -17,5 +17,7 @@ export class Utils {
         status: 45,
 
         flag: 'assets/user-card-icons/canada-flag.svg',
+
```
Comment (tatianahutii) [inline frontend/src/app/shared/utils/user-card.utils.ts:20]: I believe we can delete this helper

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by tatianahutii — APPROVED** [review]

**Review by tatianahutii — APPROVED** [review]

**Review by Anatolii-Ivanov — APPROVED** [review]
